jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java

```diff
@@ -0,0 +1,141 @@
+            } else {
+                classMembers = getClassMembersCount(v -> v.doubleValue() >= current
+                        && v.doubleValue() < next, data);
+            }
+            percentages[i] = (classMembers / totalSize) * 100;
+        }
+        return percentages;
+    }
+
+
+    private double getClassMembersCount(Predicate<Double> predicate, List<Double> values) {
+        int classMembers = 0;
+        for (int i = 0; i < values.size(); i++) {
+            if (predicate.test(values.get(i))) {
+                classMembers++;
+            }
+        }
+        return classMembers;
+    }
+
+    // histogram natural breaks and quantile
+    double[] getPercentages(List<HistogramClassification.Bucket> buckets,
+                            List<Double> breaks, int numClasses) {
+        double[] percentages = new double[numClasses];
+        double[] arClassMembers = new double[numClasses];
+        int totalSize = 0;
+        for (int i = 0; i < buckets.size(); i++) {
+            HistogramClassification.Bucket b = buckets.get(i);
+            totalSize += b.getCount();
+            getClassMembersCount(numClasses, breaks, arClassMembers, b);
+        }
+        for (int i = 0; i < numClasses; i++) {
+            percentages[i] = (arClassMembers[i] / totalSize) * 100;
+        }
+        return percentages;
+    }
+
+    private void getClassMembersCount(int numClasses,
+                                      List<Double> breaks,
+                                      double[] arClassMembers,
+                                      HistogramClassification.Bucket b) {
+        int bucketCount = b.getCount();
+        for (int i = 0; i < numClasses; i++) {
+            double current = breaks.get(i);
+            double next = breaks.get(i + 1);
+            boolean last = i + 1 == numClasses;
+            Double overlapping;
+            if (last) {
+                overlapping = getOverlappingPercentages(current, next, b.getMin(), b.getMax());
+            } else {
+                overlapping = getOverlappingPercentages(current, Math.nextDown(next), b.getMin(), b.getMax());
+            }
+            if (overlapping.doubleValue() == 1.0) {
+                arClassMembers[i] += bucketCount;
+            } else if (overlapping.doubleValue() == 0.0) {
+                arClassMembers[i] += 0.0;
+            } else {
+                double classMembers = overlapping * bucketCount;
+                arClassMembers[i] += classMembers;
+            }
+        }
+    }
+
+
+    private Double getOverlappingPercentages(double cMin, double cMax,
+                                             double bMin, double bMax) {
+        if (cMin > bMax || cMax < bMin) {
+            return 0.0;
+        }
+        double regionMin = Math.max(cMin, bMin);
+        double regionMax = Math.min(cMax, bMax);
+        double dividend = regionMax - regionMin;
+        double divider = bMax - bMin;
+        if (dividend == divider)
```

Comment:
But why not check explicitly for zero? This is opaque, it seems like you're trying to optimize away a division.

Reply:
fixed